src/Microsoft.AspNetCore.Mvc.Razor/TagHelpers/UrlResolutionTagHelper.cs

```diff
@@ -250,41 +245,104 @@ protected bool TryResolveUrl(string url, out string resolvedUrl)
+                appRelativeUrl,
+                appRelativeUrl.Length - postTildeSlashUrlValue.Length,
+                postTildeSlashUrlValue);
+
+            return true;
+        }
+
+        private static int FindRelativeStart(string url)
+        {
+            if (url == null || url.Length < 2)
+            {
+                return -1;
+            }
+
+            var maxTestLength = url.Length - 2;
+
+            var start = 0;
+            for (; start < url.Length; start++)
+            {
+                if (start > maxTestLength)
+                {
+                    return -1;
+                }
+
+                if (!IsCharWhitespace(url[start]))
+                {
+                    break;
+                }
+            }
 
             // Before doing more work, ensure that the URL we're looking at is app-relative.
-            if (trimmedUrl.Length >= 2 && trimmedUrl[0] == '~' && trimmedUrl[1] == '/')
+            if (url[start] != '~' || url[start + 1] != '/')
             {
-                var urlHelper = UrlHelperFactory.GetUrlHelper(ViewContext);
-                var appRelativeUrl = urlHelper.Content(trimmedUrl);
-                var postTildeSlashUrlValue = trimmedUrl.Substring(2);
+                return -1;
+            }
 
-                if (!appRelativeUrl.EndsWith(postTildeSlashUrlValue, StringComparison.Ordinal))
+            return start;
+        }
+
+        private static string CreateTrimmedString(string input, int start)
+        {
+            var end = input.Length - 1;
+            for (; end >= start; end--)
+            {
+                if (!IsCharWhitespace(input[end]))
                 {
-                    throw new InvalidOperationException(
-                        Resources.FormatCouldNotResolveApplicationRelativeUrl_TagHelper(
-                            url,
-                            nameof(IUrlHelper),
-                            nameof(IUrlHelper.Content),
-                            "removeTagHelper",
-                            typeof(UrlResolutionTagHelper).FullName,
-                            typeof(UrlResolutionTagHelper).GetTypeInfo().Assembly.GetName().Name));
+                    break;
                 }
+            }
 
-                resolvedUrl = new EncodeFirstSegmentContent(
-                    appRelativeUrl,
-                    appRelativeUrl.Length - postTildeSlashUrlValue.Length,
-                    postTildeSlashUrlValue);
+            var len = end - start + 1;
 
-                return true;
+            // Substring returns same string if start == 0 && len == Length
+            return input.Substring(start, len);
+        }
+
+        private static bool IsCharWhitespace(char ch)
+        {
+            var i = 0;
+            for (; i < ValidAttributeWhitespaceChars.Length; i++)
+            {
+                if (ValidAttributeWhitespaceChars[i] == ch)
+                {
+                    break;
```

Comment:
`return true;` and you should be able to remove the outer `if` with a `return false`